File: ios/App/App/ThredMobileCore.swift
//
//  thred-message.swift
//  App
//
//  Created by Arta Koroushnia on 2023-02-05.
//

import Capacitor

@objc(ThredMobileCore)
public class ThredMobileCore: CAPPlugin {
    
    var appVC: AppController?
    
    func initController(storyboard: String, vc: String) -> UIViewController?{
        let storyboard = UIStoryboard(name: storyboard, bundle: nil)
        return storyboard.instantiateViewController(withIdentifier: vc)
    }
    
    var handlers = [String : ((Any) -> ())]()
    
    
    @objc func sendResponse(_ call: CAPPluginCall) {
        if let id = call.getString("id"), let data = call.getString("data"), let replyHandler = handlers[id]{
            print(id)
            print(data)
            replyHandler(data)
            handlers[id] = nil
        }
    }
    
    @objc func toggleAppVisible(_ call: CAPPluginCall) {
        print("toggling")

        DispatchQueue.main.async {
            if let toggle = call.getBool("visible"), let appVC = self.appVC, let vc = self.bridge?.viewController as? CAPBridgeViewController{
                if toggle{
                    vc.present(appVC, animated: false, completion: nil)
                }
                else{
                    appVC.dismiss(animated: false)
                }
            }
        }
    }
    
    func confirmTransaction(txData: String, replyHandler: @escaping (Any) -> ()){
            
            let uuid = UUID().uuidString
            
            self.handlers[uuid] = replyHandler
                    
            self.notifyListeners("newTransaction", data: ["request": txData, "id": uuid])
                    
            
        
    }
    
    
    @objc func setWallet(_ call: CAPPluginCall) {
        
    }
    
    
    @objc func getWallet(_ call: CAPPluginCall) {
          
    }
    
    @objc func openApp(_ call: CAPPluginCall) {
        print("open app")
        DispatchQueue.main.async{
            if let data = call.getObject("app"), let signer = call.getString("signer"), let vc = self.bridge?.viewController as? CAPBridgeViewController, let appVC = self.initController(storyboard: "Main", vc: "app") as? AppController{
                self.appVC = appVC
                print(data)
                print(signer)
                appVC.app = data
                appVC.injectedSigner = signer
                appVC.thredCore = self
                vc.present(appVC, animated: true, completion: nil)
                
            }
        }
    }
    
}

File: android/app/src/main/java/com/thredwallet/app/thred/ThredMobileCore.kt
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.*

@CapacitorPlugin(name = "ThredMobileCore")
class ThredMobileCore : Plugin() {

  val handlers = mutableMapOf<String, (result: Any?) -> Unit>()

  @PluginMethod
  fun sendResponse(call: PluginCall) {

    val id = call.getString("id")
    val data = call.getString("data")

    val handler = handlers[id]

    if (handler != null){
      handler(data)
      handlers.remove(id)
    }
  }

  @PluginMethod
  fun openApp(call: PluginCall) {

    val signer = call.getString("signer")
    val data: JSObject = call.getObject("app")
    val margins: JSObject = call.getObject("margins")
    val app = data.getString("loadURL")

    //open app controller


    if (app != null && signer != null){
      var topMargin = margins.getInteger("top")
      var bottomMargin = margins.getInteger("bottom")

      if (topMargin == null){
        topMargin = 0
      }
      if (bottomMargin == null){
        bottomMargin = 0
      }

      val appMargins = mutableMapOf<String, Int>()
      appMargins["bottom"] = bottomMargin
      appMargins["top"] = topMargin

      TBridgeActivity.getInstanceActivity()?.toggleAppVisibility(1, app, signer, appMargins)
    }
  }

  @PluginMethod
  fun setWallet(call: PluginCall?) {
  }

  @PluginMethod
  fun getWallet(call: PluginCall?) {
  }

  @PluginMethod
  fun toggleAppVisible(call: PluginCall) {
    val show = call.getBoolean("visible")
    val toggle: Int?
    if (show != null){
      toggle = if (show){
        3
      } else{
        4
      }
      TBridgeActivity.getInstanceActivity()?.toggleAppVisibility(toggle, null, null, null)
    }
  }

  fun confirmTransaction(txData: String, replyHandler: (result: Any?) -> Unit){

    val uuid = UUID.randomUUID().toString()

    handlers[uuid] = replyHandler

    val ret = JSObject()
    ret.put("request", txData)
    ret.put("id", uuid)

    notifyListeners("newTransaction", ret)
  }
}
